tools/flow_grid.py

```
@@ -0,0 +1,689 @@
+                     representing the gridded data. Default values are used
+                     internally by some class methods:
+                         'dem' : digital elevation data
+                         'dir' : flow direction data
+                         'acc' : flow accumulation (upstream area) data
+                    
+        input_type : 'raster', 'ascii' or 'array'
+                     Type of input data.
+        band : int
+               For raster data, the band number to read.
+        nodata : int or float
+                 Value indicating no data.
+        bbox : tuple or list
+               Bounding box, if none provided.
+
+        """
+
+        # read ascii file
+        if input_type == 'ascii':
+            with open(data) as header:
+                ncols = int(header.readline().split()[1])
+                nrows = int(header.readline().split()[1])
+                xll = ast.literal_eval(header.readline().split()[1])
+                yll = ast.literal_eval(header.readline().split()[1])
+                cellsize = ast.literal_eval(header.readline().split()[1])
+                nodata = ast.literal_eval(header.readline().split()[1])
+                shape = (nrows, ncols)
+                bbox = (xll, yll, xll + ncols*cellsize, yll + nrows*cellsize)
+            data = np.loadtxt(data, skiprows=6, **kwargs)
+            nodata = data.dtype.type(nodata)
+
+        # read raster file
+        if input_type == 'raster':
+            import rasterio
+            f = rasterio.open(data)
+            crs = f.crs
+            bbox = tuple(f.bounds)
+            shape = f.shape
+            cellsize = f.affine[0]
+            nodata = f.nodatavals[0]
+            if len(f.indexes) > 1:
+                data = np.ma.filled(f.read_band(band))
+            else:
+                data = np.ma.filled(f.read())
+                f.close()
+                data = data.reshape(shape)
+            nodata = data.dtype.type(nodata)
+
```

what happens if `input_type` is not in `['ascii', 'raster', 'array']?  Should we raise an error.  It looks like we just continue on.  Let's use an `if/elif/else` series to test for the input type with the `else` raising an error.